dlnd/dlnd_datamode_rof_forcing_mod.F90

```diff
@@ -0,0 +1,212 @@
+
+   public :: dlnd_datamode_rof_forcing_advertise
+   public :: dlnd_datamode_rof_forcing_init_pointers
+   public :: dlnd_datamode_rof_forcing_advance
+
+   ! module pointer arrays
+   real(r8), pointer :: lfrac(:)
+
+   character(*), parameter :: Flrl_rofsur_nonh2o = 'Flrl_rofsur_nonh2o'
+   character(*), parameter :: Flrl_rofsur        = 'Flrl_rofsur'
+   character(*), parameter :: Flrl_rofsub        = 'Flrl_rofsub'
+   character(*), parameter :: Flrl_rofgwl        = 'Flrl_rofgwl'
+   character(*), parameter :: Flrl_rofi          = 'Flrl_rofi'
+   character(*), parameter :: Flrl_irrig         = 'Flrl_irrig'
+
+   character(len=11) :: fldnames_h2o(5) = &
+        (/'Flrl_rofsur', 'Flrl_rofsub','Flrl_rofgwl','Flrl_rofi  ','Flrl_irrig '/)
+
+   integer :: ntracers_nonh2o
+
+   character(*), parameter :: nullstr = 'null'
+   character(*), parameter :: u_FILE_u = &
+        __FILE__
+
+!===============================================================================
+contains
+!===============================================================================
+
+   subroutine dlnd_datamode_rof_forcing_advertise(exportState, fldsExport, flds_scalar_name, logunit, mainproc, rc)
+
+      ! determine export fields to advertise to mediator
+
+      ! input/output arguments
+      type(ESMF_State)                 :: exportState
+      type(fldList_type) , pointer     :: fldsExport
+      character(len=*)   , intent(in)  :: flds_scalar_name
+      integer            , intent(in)  :: logunit
+      logical            , intent(in)  :: mainproc
+      integer            , intent(out) :: rc
+
+      ! local variables
+      character(len=CS) :: lnd2rof_tracers
+      type(fldlist_type), pointer :: fldList
+      !-------------------------------------------------------------------------------
+
+      rc = ESMF_SUCCESS
+
+      ! lnd2rof liquid tracers (liquid tracers OTHER than water)
+      ! coupling the land input of tracers other than standard water to MOSART
+      if (mainproc) then
+         write(logunit,'(a)') 'reading in non-water tracers from land (if any) in drv_flds_in '
+      end if
+      lnd2rof_tracers = ' '
+      call shr_lnd2rof_tracers_readnl('drv_flds_in', lnd2rof_tracers)
+      if (lnd2rof_tracers /= ' ') then
+         ntracers_nonh2o = shr_string_listGetNum(lnd2rof_tracers)
+      else
+         ntracers_nonh2o = 0
+      end if
+
+      !-------------------
+      ! Advertise export fields to rof
+      !-------------------
+
+      call dshr_fldList_add(fldsExport, trim(flds_scalar_name))
+      call dshr_fldlist_add(fldsExport, "Sl_lfrin")
+
+      ! The following puts all non-water tracers as an undidstributed dimension in the export state field
+      if (ntracers_nonh2o > 1) then
+         call dshr_fldList_add(fldsExport, Flrl_rofsur_nonh2o, ungridded_lbound=1, ungridded_ubound=ntracers_nonh2o)
+      else if (ntracers_nonh2o == 1) then
+         call dshr_fldList_add(fldsExport, Flrl_rofsur_nonh2o)
+      end if
```

Comment:
It seems confusing to me that you skip having an ungridded dimension if you have a single tracer. Why not have an ungridded dimension with size 1 in this case so you can avoid special handling? (I see that there may still be a need for special handling of the case where there are 0 tracers, in which case you won't have this field at all.)

Reply:
We'll address this later.